Modules/_collectionsmodule.c

```diff
@@ -537,12 +538,24 @@ deque_copy(PyObject *deque)
         Py_DECREF(new_deque);
         return NULL;
     }
-    if (old_deque->maxlen < 0)
-        return PyObject_CallFunctionObjArgs((PyObject *)(Py_TYPE(deque)),
-                                            deque, NULL);
-    else
-        return PyObject_CallFunction((PyObject *)(Py_TYPE(deque)), "Oi",
-            deque, old_deque->maxlen, NULL);
+    if (old_deque->maxlen < 0) {
+        result = PyObject_CallFunctionObjArgs((PyObject *)(Py_TYPE(deque)),
+                                              deque, NULL);
+    }
+    else {
+        result = PyObject_CallFunction((PyObject *)(Py_TYPE(deque)), "Oi",
+                                       deque, old_deque->maxlen, NULL);
+    }
+    if (result != NULL &&
+        !PyObject_IsInstance(result, (PyObject *)&deque_type))
+    {
+        PyErr_Format(PyExc_TypeError,
+                     "%.200s.__new__() must return a deque, not %.200s",
```

Comment:
Bad `__new__()` is not the only possible cause. The type can be an instance of a metaclass with overridden `__call__`.